src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+{
+	scmp_filter_ctx ctx;
+	size_t i;
+	int rc;
+
+	/* initialize the libseccomp context */
+	ctx = seccomp_init(SCMP_ACT_KILL);
+	if (ctx == NULL) {
+		fprintf(stderr, "Could not initialize the sandbox (seccomp_init failed)\n");
+		exit(1);
+	}
+
+	/* syscalls without argument filtering */
+	const int noarg_whitelist[20] = {
+		SCMP_SYS(brk),
+		SCMP_SYS(chdir),
+		SCMP_SYS(clone),
+		SCMP_SYS(close),
+		SCMP_SYS(exit_group),
+		SCMP_SYS(fstat),
+		SCMP_SYS(getegid),
+		SCMP_SYS(geteuid),
+		SCMP_SYS(getgid),
+		SCMP_SYS(getpid),
+		SCMP_SYS(getuid),
+		SCMP_SYS(lseek),
+		SCMP_SYS(munmap),
+		SCMP_SYS(pipe),
+		SCMP_SYS(rt_sigreturn),
+		SCMP_SYS(set_tid_address),
+		SCMP_SYS(stat),
+		SCMP_SYS(umask),
+		SCMP_SYS(uname)
+	};
+	for (i = 0; i < 20; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, noarg_whitelist[i], 0);
+		if (rc < 0)
+			goto fail;
+
+		rc = seccomp_syscall_priority(ctx, noarg_whitelist[i], 100);
+		if (rc < 0)
+			goto fail;
+	}
+
+	/* mkdir */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mkdir), 1,
+		SCMP_A1(SCMP_CMP_EQ, 0700)); /* src/core/settings.c:828 */
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(mkdir), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* setsockopt */
+	struct setsockopt_struct_t {
+		int sockfd;
+		int level;
+		int optname;
+		socklen_t optlen;
+	};
+	const struct setsockopt_struct_t setsockopt_struct[2] = {
+		{ 4, SOL_SOCKET, SO_KEEPALIVE, 4 },
+		{ 4, SOL_SOCKET, SO_REUSEADDR, 4 }
+	};
+	for (i = 0; i < 2; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(setsockopt), 4,
+			SCMP_A0(SCMP_CMP_EQ, setsockopt_struct[i].sockfd),
+			SCMP_A1(SCMP_CMP_EQ, setsockopt_struct[i].level),
+			SCMP_A2(SCMP_CMP_EQ, setsockopt_struct[i].optname),
+			SCMP_A4(SCMP_CMP_EQ, setsockopt_struct[i].optlen));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(setsockopt), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* getsockopt */
+	struct getsockopt_struct_t {
+		int sockfd;
+		int level;
+		int optname;
+	};
+	const struct getsockopt_struct_t getsockopt_struct[2] = {
+		{ 4, SOL_SOCKET, SO_TYPE  },
+		{ 4, SOL_SOCKET, SO_ERROR }
+	};
+	for (i = 0; i < 2; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(getsockopt), 3,
+			SCMP_A0(SCMP_CMP_EQ, getsockopt_struct[i].sockfd),
+			SCMP_A1(SCMP_CMP_EQ, getsockopt_struct[i].level),
+			SCMP_A2(SCMP_CMP_EQ, getsockopt_struct[i].optname));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(getsockopt), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* ioctl */
+	const int ioctl_array[3] = {
+		TCSETSW,
+		TCGETS,
+		TIOCGWINSZ
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(ioctl), 1,
+			SCMP_A1(SCMP_CMP_EQ, ioctl_array[i]));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(ioctl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* kill */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(kill), 1,
```

Comment:
The ioctls may change and we may need to adjust them in the future, so putting it in an array makes it much easier to add support to new ones. The kill array comes from hardcoded signals which are not likely to change, so putting them in an array would only serve to needlessly increase the size of the code.